pkg/registry/core/service/storage/rest.go

```diff
@@ -226,10 +226,7 @@ func (rs *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 	nodePortOp := portallocator.StartOperation(rs.serviceNodePorts, dryrun.IsDryRun(options.DryRun))
 	defer nodePortOp.Finish()
 
-	// TODO: This creates nodePorts if needed. In the future nodePorts may be cleared if *never* used.
-	// But for now we stick to the KEP "don't allocate new node ports but do not deallocate existing node ports if set"
-	if service.Spec.Type == api.ServiceTypeNodePort ||
-		(service.Spec.Type == api.ServiceTypeLoadBalancer && shouldAllocateNodePorts(service)) {
+	if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer {
```

Comment:
Does this mean this is an existing bug that applied for any service type that isn't NodePort or LoadBalancer? 

I _think_ it's valid to specify a nodePort for a ClusterIP type Service for example, are those node port allocations not considered either?

Reply:
Unless the gate was enabled AND the field was changed from true (default) to false, it should work correctly.  Only when the field is false does it get this wrong.

Reply:
Ah I see, so setting `nodePort: 30000` on a `ClusterIP` type Service is still accounted for

Reply:
... but it will fail in validation
https://github.com/kubernetes/kubernetes/blob/d7123a65248e25b86018ba8220b671cd483d6797/pkg/apis/core/validation/validation.go#L4404-L4412

Reply:
Ah I missed that part, thanks :)